Escape string for JSON output(--json)

yu74n opened the pull request:
Normally, `redis-cli` escapes non-printable data received from Redis, using a custom scheme (which is also used to handle quoted input). When using `--json` this is not desired as it is not compatible with RFC 7159, which specifies JSON strings are assumed to be Unicode and how they should be escaped.

This commit changes `--json` to follow RFC 7159, which means that properly encoded Unicode strings in Redis will result with a valid Unicode JSON.

However, this introduces a new problem with `--json` and data that is not valid Unicode (e.g., random binary data, text that follows other encoding, etc.). To address this, we add `--quoted-json` which produces JSON strings that follow the original redis-cli quoting scheme.

For example, a value that consists of only null (0x00) bytes will show up as:
* `"\u0000\u0000\u0000"` when using `--json`
* `"\\x00\\x00\\x00"` when using `--quoted-json`

oranagra reviewed (commented):
@yu74n seems like your new test is failing.
also, i don't think you should use hi_sds...  directly. the rest of the code sds defines from sdscompat.h

besides, can you please list the differences between sdscatrepr and your modified version?

as far as i can tell, `\f` is probably already implicitly handled by `isprint` (not sure why `\a` and `\b` are there either).
so the only changes i can spot is the escaping of `/` (not sure why it's needed),
and the double `\\` in the `isprint` else which seem like a bug, so please tell me what i'm missing.

yu74n commented:
Fixed to use `sds` instead of `hi_sds`.

The differences are
|----|sdscatrepr|escapeJSON|note|
|----|----|----|----|
|not printable char(0x0-0x19, 0x80-)| with one reverse solidus(\xXX) | with two reverse solidus(\\\\xXX) |----|
| '/' | not escaped | escaped |----|
| \a, \b| human readable | \xXX style, \a(0x7), \b(0x8) | both human readable(\a) or \xXX are okay |

\a, \b, \f are printed as \xXX is no problem, then change to not convert.
About '/',  according to `jq`, both escaped and not escaped seems fine.

yu74n commented:
About unit test, actually the new tests I added are passed in my local environment, I try to reproduce this error, but not yet.
I'll try to run them in ubuntu and other environments tomorrow.

yu74n commented:
@oranagra when a user call SET with unicode character, encoded character like UTF-8 is set, and GET returns just encoded string.
For example, in the C Clef character(U+1D11E) case, 
```
> set test 𝄞
> get test
"\xf0\x9d\x84\x9e"
```
0xF0 0x9D 0x84 0x9E is by UTF-8 encoding, JSON parser just handle it with string, so "\xf0" should be "\\xf0".

```
$ echo '{"test": "\xf0"}' | jq
parse error: Invalid escape at line 1, column 15
$ echo '{"test": "\\xf0"}' | jq
{
  "test": "\\xf0"
}
```

yu74n commented:
Fixed the failed test, I used '𝄞' (U+1D11E) for test, it was encoded to "0xF0 0x9D 0x84 0x9E" with UTF-8.
I guess some test environments using ISO-8859-1, in this env, 0xF0 is decoded to 'ð' and others to non-printable character, for example 0x9D is to `<OSC>`, that's why my tests were failed.
Then I fixed to use 0xF0 to non printable character test.

yossigo commented:
@yu74n @oranagra
I'm not sure this form of escaping is correct, although I agree there's a problem here. In the example above, `redis-cli` applies its own custom encoding scheme to represent the series of binary bytes. However, JSON does not follow this scheme and the RFC clearly states that:

> All Unicode characters may be placed within the
   quotation marks, except for the characters that must be escaped:
   quotation mark, reverse solidus, and the control characters (U+0000
   through U+001F).

So, I think a more correct way to handle `--json` is to actually avoid escaping altogether (with the exceptions mentioned in the RFC) and produce the binary output which, in the context of JSON, will be decoded as Unicode.

yu74n commented:
@yossigo @oranagra 
I worried that there are cases to store binary data like serialized Java Object. 
In this case, I'm wondering if printing it with unicode is useful or not.
I googled how the best way to handle binary data in JSON, it suggests using Base64, however, in Redis, it's hard to determine the data type is binary data or String since binary data is almost just store as String type.

yossigo commented:
@yu74n I agree with you.

The only possible solution I see in these cases is to add a new `--json-encoding <encoding>` option which will let the user decide if non-printable data should be printed as-is (suitable for Unicode strings), or encoded in different ways (i.e. current `redis-cli` scheme, base64, etc.).

yu74n commented:
I realized RFC said
> unescaped = %x20-21 / %x23-5B / %x5D-10FFFF

then we have to care about 0x22("), 0x5c(\) and some special characters(backspace, form feed, line feed, carriage return and tab). About the characters in 5D~10FFFF, we can print as is, I think.

```
Set 🆎 (U+1F18E)
$ redis-cli hset ab 🆎 🆎

Without --json, escape the output
$ redis-cli hgetall ab
1) "\xf0\x9f\x86\x8e"
2) "\xf0\x9f\x86\x8e"

With --json, don't have to escape
$ redis-cli --json hgetall ab
{"🆎":"🆎"}
```

For binary data, it's better to provide an option as @yossigo said, otherwise binary data is represent as raw data with `--json`.
```
src/redis-cli hget hg:image image1 
"\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\
...
(omitted)

redis-cli --json hget hg:image image1
"����JFIFHH 
...
(omitted)

```

yossigo commented:
@yu74n I've pushed a commit with a couple of minor changes, please let me know if that makes sense.

yu74n commented:
@yossigo Thank you for fix 0x00-0x1f and the special character in the test.

oranagra commented on src/cli_common.c:
```diff
@@ -371,3 +371,46 @@ void freeCliConnInfo(cliConnInfo connInfo){
+
+/*
+ * Escape a string for JSON output(--json)
+ * Following characters should be escaped with '\' in JSON
```
comment is outdated?

yu74n replied:
I think these characters are still escaped with '\\'

oranagra replied:
yes, but that's not all that this function does.

oranagra commented on src/redis-cli.c:
```diff
@@ -1216,20 +1223,21 @@ static sds cliFormatReplyJson(sds out, redisReply *r) {
                 key->type == REDIS_REPLY_STRING ||
                 key->type == REDIS_REPLY_VERB) {
-                out = cliFormatReplyJson(out, key);
+                out = cliFormatReplyJson(out, key, JSON_DISABLE_BINARY_ENCODING);
```
Why are map fields always not using binary escaping? 
Maybe add a comment?

oranagra commented on src/redis-cli.c:
```diff
@@ -261,6 +263,7 @@ static struct config {
     int resp3; /* value of 1: specified explicitly, value of 2: implicit like --json option */
     int in_multi;
     int pre_multi_dbnum;
+    int json_binary_encoding;
```
This fields isn't initialized. 
Is it expected to be one of JSON_ENABLE_BINARY_ENCODING / JSON_DISABLE_BINARY_ENCODING? 
I.e. 1 or 2?
Why not make it a boolean?

yu74n replied:
I thought that other options like BASE64 format, but I don't think I want to implement them in this time because I'm not sure how much the feature will be useful currently, then I'll fix it as boolean.

oranagra replied:
don't forget to init it in the beginning of `main()`

oranagra commented on src/redis-cli.c:
```diff
@@ -1194,7 +1197,11 @@ static sds cliFormatReplyJson(sds out, redisReply *r) {
     case REDIS_REPLY_STRING:
     case REDIS_REPLY_VERB:
-        out = sdscatrepr(out,r->str,r->len);
+        if (binary == 1) {
```
Why use `== 1` either match it against JSON_ENABLE_BINARY_ENCODING, or just use it as a boolean.

oranagra commented on src/redis-cli.c:
```diff
@@ -1216,20 +1223,21 @@ static sds cliFormatReplyJson(sds out, redisReply *r) {
-                sds tmp = cliFormatReplyJson(sdsempty(), key);
-                out = sdscatrepr(out,tmp,sdslen(tmp));
+                sds tmp = cliFormatReplyJson(sdsempty(), key, JSON_DISABLE_BINARY_ENCODING);
+                out = escapeJsonString(out,tmp,sdslen(tmp));
```
Isn't the output of the previous line already escaped? in which case all we need is add quotes? 
I'm not sure..

yu74n replied:
The first one is to convert non-string type to string since JSON key should be string, the second one is to escape the converted string.
I'll comment it in the source code as well.

oranagra commented on src/redis-cli.c:
```diff
@@ -1978,6 +1993,7 @@ static void usage(int err) {
 "  --quoted-input     Force input to be handled as quoted strings.\n"
 "  --csv              Output in CSV format.\n"
 "  --json             Output in JSON format (default RESP3, use -2 if you want to use with RESP2).\n"
+"  --json-binary-encoding Output JSON value as escaped binary character with \\x?? style\n"
```
Truth be told, I don't really understand the value of this. 
Maybe improve the doc, and please also update the PR top comment to list that option and explain it.

yu74n replied:
Updated the PR description.

oranagra commented on src/redis-cli.c:
```diff
@@ -1225,22 +1244,20 @@ static sds cliFormatReplyJson(sds out, redisReply *r, int binary) {
-                     * otherwise all keys are also printed as binary string, it would be useless
-                     * for users */
-                    out = cliFormatReplyJson(out, key, 0);
+                    out = cliFormatReplyJson(out,key,mode);
```
```suggestion
                out = cliFormatReplyJson(out,key,mode);
```

oranagra commented on src/redis-cli.c:
```diff
@@ -1658,8 +1675,12 @@ static int parseOptions(int argc, char **argv) {
-        } else if (!strcmp(argv[i],"--json-binary-encoding")) {
-            config.json_binary_encoding = 1;
+        } else if (!strcmp(argv[i],"--quoted-json")) {
+            /* Not overwrite explicit value by -3*/
```
```suggestion
            /* Not overwrite explicit value by -3 */
```

oranagra reviewed (approved):
LGTM

oranagra commented:
@redis/core-team please approve new interface for redis-cli (`--quoted-json`) see the top comment.

madolson reviewed (approved):
Change seems reasonable to me.

enjoy-binbin commented on tests/integration/redis-cli.tcl:
```diff
@@ -228,6 +228,30 @@ start_server {tags {"cli"}} {
+        set eth "\xf0\x65"
+        r hset eth test $eth
+        assert_equal \"\\xf0e\" [run_cli hget eth test]
+        assert_equal \"\xf0e\" [run_cli --json hget eth test]
```
https://github.com/redis/redis/runs/5435634922?check_suite_focus=true#step:5:4649
```
Expected '""' to be equal to '"ðe"' (context: type source line 241 file /Users/runner/work/redis/redis/tests/integration/redis-cli.tcl cmd {assert_equal \"\xf0e\" [run_cli --json hget eth test]} proc ::test)
```

enjoy-binbin commented on tests/integration/redis-cli.tcl:
```diff
@@ -228,6 +228,30 @@ start_server {tags {"cli"}} {
+        assert_equal \{\"1\":\"One\"\} [run_cli --json hgetall numkey]
+        # non-string, non-printable keys
+        r hset npkey "K\x00\x01ey" "V\x00\x01alue"
+        assert_equal \{\"K\\u0000\\u0001ey\":\"V\\u0000\\u0001alue\"\} [run_cli --json hgetall npkey]
```
*** [err]: Non-interactive TTY CLI: Escape character in JSON mode in tests/integration/redis-cli.tcl
Expected '{"K\u0000\u0001ey":"V\u0000\u0001alue"}' to be equal to '{"K\u0000\u001ey":"V\u0000\u001alue"}' (context: type eval line 21 cmd {assert_equal \{\"K\\u0000\\u0001ey\":\"V\\u0000\\u0001alue\"\} [run_cli --json hgetall npkey]} proc ::test)

enjoy-binbin commented on tests/integration/redis-cli.tcl:
```diff
@@ -228,6 +228,30 @@ start_server {tags {"cli"}} {
+        # non-string, non-printable keys
+        r hset npkey "K\x00\x01ey" "V\x00\x01alue"
+        assert_equal \{\"K\\u0000\\u0001ey\":\"V\\u0000\\u0001alue\"\} [run_cli --json hgetall npkey]
+        assert_equal \{\"K\\\\x00\\\\x01ey\":\"V\\\\x00\\\\x01alue\"\} [run_cli --quoted-json hgetall npkey]
```
*** [err]: Non-interactive TTY CLI: Escape character in JSON mode in tests/integration/redis-cli.tcl
Expected '{"K\\x00\\x01ey":"V\\x00\\x01alue"}' to be equal to '{"K\\x00\\x1ey":"V\\x00\\x1alue"}' (context: type eval line 22 cmd {assert_equal \{\"K\\\\x00\\\\x01ey\":\"V\\\\x00\\\\x01alue\"\} [run_cli --quoted-json hgetall npkey]} proc ::test)

enjoy-binbin reviewed (commented):
these tests will fail on my machine(centos7, gcc 9.3.1)

yossigo commented:
@enjoy-binbin Thanks, seems like a TCL version issue. Hopefully fixed by #10386.
